typescript/infra/src/govern/HyperlaneHaasGovernor.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  ChainName,
+  CheckerViolation,
+  HyperlaneCoreChecker,
+  InterchainAccount,
+  InterchainAccountChecker,
+} from '@hyperlane-xyz/sdk';
+
+import {
+  AnnotatedCallData,
+  HyperlaneAppGovernor,
+} from './HyperlaneAppGovernor.js';
+import { HyperlaneCoreGovernor } from './HyperlaneCoreGovernor.js';
+import { ProxiedRouterGovernor } from './ProxiedRouterGovernor.js';
+
+export class HyperlaneHaasGovernor extends HyperlaneAppGovernor<any, any> {
+  protected readonly icaGovernor: ProxiedRouterGovernor<any, any>;
+  protected readonly coreGovernor: HyperlaneCoreGovernor;
+
+  constructor(
+    ica: InterchainAccount,
+    private readonly icaChecker: InterchainAccountChecker,
+    private readonly coreChecker: HyperlaneCoreChecker,
+  ) {
+    super(coreChecker, ica);
+
+    this.icaGovernor = new ProxiedRouterGovernor(this.icaChecker);
+    this.coreGovernor = new HyperlaneCoreGovernor(this.coreChecker, this.ica);
+  }
+
+  protected mapViolationToCall(
+    _: CheckerViolation,
+  ): Promise<{ chain: string; call: AnnotatedCallData } | undefined> {
+    throw new Error(`HyperlaneHaasGovernor has no native map of violations.`);
+  }
+
+  // Handle ICA violations before Core violations
+  protected async mapViolationsToCalls(): Promise<void> {
+    // Handle ICA violations first
+    const icaCallObjs = await Promise.all(
+      this.icaChecker.violations.map((violation) =>
+        this.icaGovernor.mapViolationToCall(violation),
+      ),
+    );
+
+    // Process ICA call objects
+    for (const callObj of icaCallObjs) {
+      if (callObj) {
+        this.pushCall(callObj.chain, callObj.call);
+      }
+    }
+
+    // Then handle Core violations
+    const coreCallObjs = await Promise.all(
+      this.coreChecker.violations.map((violation) =>
+        this.coreGovernor.mapViolationToCall(violation),
+      ),
+    );
+
+    // Process Core call objects
+    for (const callObj of coreCallObjs) {
+      if (callObj) {
+        this.pushCall(callObj.chain, callObj.call);
+      }
+    }
+  }
+
+  async check() {
+    await this.icaChecker.check();
+    await this.coreChecker.check();
+  }
+
+  async checkChain(chain: ChainName) {
+    await this.icaChecker.checkChain(chain);
+    await this.coreChecker.checkChain(chain);
+  }
+
+  getCheckerViolations() {
+    return [...this.icaChecker.violations, ...this.coreChecker.violations];
+  }
+
+  async govern(confirm = true, chain?: ChainName) {
```

Comment:
didn't you say some filtering for dupes was needed? not sure i see it

Reply:
yup, this was added to the base class https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/4435/files#diff-07e7a8ab698661948a9707a204e898b70a3bba0c6d1fe64d8d633bb968ee5bb0R186-R194